pkg/kubecost/allocation.go

```diff
@@ -2261,7 +2261,71 @@ func (asr *AllocationSetRange) Accumulate() (*AllocationSet, error) {
 }
 
 // TODO accumulate into lower-resolution chunks of the given resolution
-// func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) *AllocationSetRange
+func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) (*AllocationSetRange, error) {
+	allocSetRange := &AllocationSetRange{allocations: []*AllocationSet{}}
+	var allocSet *AllocationSet
+	var err error
+
+	asr.Lock()
+	defer asr.Unlock()
+
+	// use window() with no lock
+	if asr.window().IsEmpty() {
+		return asr, nil
+	}
+
+	// Call total accumulate func if resolution is greater than total window duration
+	if resolution > asr.window().Duration() {
+		// unable to acquire lock here if old asr.accumulate is called
+		for _, as := range asr.allocations {
+			allocSet, err = allocSet.accumulate(as)
+			if err != nil {
+				return nil, err
+			}
+		}
+		allocSetRange.allocations = append(allocSetRange.allocations, allocSet)
+		return allocSetRange, nil
+	}
+
+	// two ways to get end of asr window
+	// 1. check if last set of asr
+	// 2. check if set window.end is asr.window.end
+	// both require no lock to be present
+	// option 1: 2 asr.window methods
+	// option 2: 2 window.length methods
+	asrWindow := asr.window()
+	asrResolution := asrWindow.Duration()
+
+	// if asrResolution < 1 hour then there is not enough data to accumulate
+	// if as.allocations(end - start) > resolution then no need, it's already accumulated. Requery to make more granular in the future?
+	// check this again?
+	if asrResolution <= time.Hour || asrResolution < resolution {
+		return asr, nil
+	}
+
+	// do not compound
+	// check as.window and accumulate till windowSum == time.duration wanted -> add accumulated set to allocSetRange
+	var currAccumulatedSum time.Duration
+	for _, as := range asr.allocations {
+		allocSet, err = allocSet.accumulate(as)
+		if err != nil {
+			return nil, err
+		}
+
+		if allocSet != nil {
+			currAccumulatedSum += allocSet.Window.Duration()
+
+			// check if end of asr to sum the final set too
+			if currAccumulatedSum >= resolution || NewWindow(nil, asrWindow.end).Equal(NewWindow(nil, allocSet.Window.end)) {
```

Comment:
What happens if I `AccumulateBy` a `resolution` that doesn't divide nicely? For example:

```
asr = [
    AS1 // 1d Window
    AS2 // 1d Window
    AS3 // 1d Window
]
```

What happens if I AccumulateBy 36 hours? I'm pretty sure the code will treat this as equivalent to calling AccumulateBy with a 2d duration because of the sum and >= logic here.

I think this is worth discussing (unless you already have!) and then documenting whatever is decided upon in a comment on `AccumulateBy` so IDEs can pick it up and show it to package users. Then we can also add a test for it.

Reply:
Great question, this was partially discussed with @nikovacevic in regards to how you would handle the (in my opinion) similar situation of asking for a resolution more granular than what is available. 

E.g. Requesting 30min resolutions on sets that are split into 1h windows:
To achieve such a granularity (at least accurately) would mean having to do a new query that is suitable for such a resolution, something I don't think is worth implementing given the required changes. I think it would be a nice to have in the future though

Reply:
> in regards to how you would handle the (in my opinion) similar situation of asking for a resolution more granular than what is available.
> E.g. Requesting 30min resolutions on sets that are split into 1h windows

Totally. I think you've made a good choice here. AFAICT the only other option is to just throw an error.

> To achieve such a granularity (at least accurately) would mean having to build a new query with parameters from the old that is suitable for such a resolution, something I don't think is worth implementing given the required changes. I think it would be a nice-to-have in the future though

That would be super cool, but probably